### ISSUE TITLE
Future direction for Decimal, embracing fixed point?

### ISSUE THREAD

**Issue by orlp** [pola-rs/polars#19784]
TL;DR: I propose to rename `Decimal` to `Decimal128`, completely drop the concept of 'precision' and only have a scale. It is always a fixed-point signed integer with value `n * 10^-scale`. Operations do not change scales, for binary operators the output scale is always the maximum of the two input scales.

**If you have a use case for decimal types, please provide feedback if you would prefer a SQL-style Decimal with a floating scale propagated by operations, or an always fixed-point Decimal where the scale doesn't float.**

---

The current design for `Decimal` is somewhat ad-hoc copied from SQL/Arrow without much careful design. In SQL you typically have a precision (total number of digits) and scale (the number of digits after the period). The idea behind operations in this Decimal design is that they compute a conservative estimate of the total precision/scale needed to represent the result, and then choose a physical data representation for the result. This also leads to a [really complicated relationship](https://learn.microsoft.com/en-us/sql/t-sql/data-types/precision-scale-and-length-transact-sql?view=sql-server-ver16) between output types and input types:

<img width="880" alt="image" src="https://github.com/user-attachments/assets/b060ed09-d6e9-44ae-aff5-c42dbcb9f065">

The scale constantly moving around is the most egregious problem. This is ideal if there were no limits involved, but in reality there are. Multiplying two decimals with each other with scale `20` would already create problems, as the resulting scale is larger than the available precision:

```python
x = pl.Series(["1.5"], dtype=pl.Decimal(scale=20))
print(x * x)
```

```
thread '<unnamed>' panicked at crates/polars-arrow/src/compute/decimal.rs:194:18:
index out of bounds: the len is 39 but the index is 40
Traceback (most recent call last):
  File "/Users/orlp/.localpython/lib/python3.11/site-packages/ptpython/repl.py", line 333, in _format_result_output
    result_repr = repr(result)
                  ^^^^^^^^^^^^
  File "/Users/orlp/.localpython/lib/python3.11/site-packages/polars/series/series.py", line 650, in __repr__
    return self.__str__()
           ^^^^^^^^^^^^^^
  File "/Users/orlp/.localpython/lib/python3.11/site-packages/polars/series/series.py", line 646, in __str__
    s_repr: str = self._s.as_str()
                  ^^^^^^^^^^^^^^^^
pyo3_runtime.PanicException: index out of bounds: the len is 39 but the index is 40
```

In duckdb:
```sql
D CREATE TABLE bar (x DECIMAL(38, 20));
D INSERT INTO bar VALUES (1.0);
D SELECT x * x FROM bar;
Out of Range Error: Needed scale 40 to accurately represent the multiplication result,
but this is out of range of the DECIMAL type. Max scale is 38; could not perform an
accurate multiplication. Either add a cast to DOUBLE, or add an explicit cast to a decimal
with a lower scale.
```

Polars straight up chokes on the problem, duckdb offers two suggestions neither of which actually solves the problem (both suggestions lose precision *before* the multiplication occurs, but I would venture to guess that most people want the multiplication output with the same number of digits after the period as the inputs).

---

What I believe most people actually want when reaching for a decimal type is a fixed-point type that has a set number of decimal digits after the period. This has very predictable semantics: each operation creates a full result which is then rounded back to the set number of decimal digits. If one wants extra digits in the result one can upcast the inputs beforehand.

What I propose is that we embrace such a fixed-point type in Polars as our Decimal support. The available range of numbers is simply set by the underlying integer type instead of a `precision` field. Initially we would only offer `Decimal128` which is backed by a signed 128-bit integer, but possibly in the future we could offer smaller/larger underlying integer types. Each `DecimalW` type would only have one parameter, scale, which sets how the number is scaled as the true number represented is `n * 10^-scale`.

Scales would not float/change based on operations. Only when mixing different scales would the smaller scale get upcasted to the larger scale, otherwise scales stay the same.

**Comment by Julian-J-S** [comment]
I created many of the Decimal issues here and for me, this sounds great! 😃 

At some point you almost always lose precision even with Decimal. Just divide a column by 3... 😆 
Also the "exploding scale" when multiplying values is almost never desired.

Most of the time you have a final scale in mind the whole time and using that from the start without any adjustments in between sounds nice.

**Comment by bdice** [comment]
I support dropping the precision aspect. But it seems like changing the scale for multiplication and division (adding/subtracting scales) is necessary, otherwise you run a high risk of integer overflow. I believe it may be necessary for correctness in division as well, due to how decimals truncate under division. This is how libcudf handles the representation/operations (no precision, just scale, and adding/subtracting scale under multiplication/division), with a design that I believe was inspired by CNL.

**Comment by Julian-J-S** [comment]
> I support dropping the precision aspect. But it seems like changing the scale for multiplication and division (adding/subtracting scales) is necessary, otherwise you run a high risk of integer overflow under multiplication. I believe it may be necessary for correctness in division as well, due to how decimals truncate under division. This is how libcudf handles the representation/operations (no precision, just scale, and adding/subtracting scale under multiplication/division), with a design that I believe was inspired by CNL.

can you elaborate on that? I don't understand.

I think the new design actually improves here as shown in the example at the top because we don't have the problem of exploding scales (less space for scale, more for precision). In the end we have a fixed width type so overflow will always be possible but this is the same with all other int types as well.


> I believe it may be necessary for correctness in division as well, due to how decimals truncate under division

not sure about this. 
Imagine having 1.00USD that you want to divide by 7.
1.00 / 7 = 0,14285714... 
There is no "correct" decimal value here. How many decimal places  do you want? 2? 4? 10? 
Choosing a "correct" amount is impossible here because it depends on the use case.
IMO keeping the original scale (=2) is great because it is predictable and if you need more you can always cast before the division to a higher scale.

**Comment by Julian-J-S** [comment]
Interesting read 🤓 

> The behavior I would propose is that division has scale `s1 - s2`

Does this mean:
- 0.1234 / 0.1 = 1.234 (scale=3; 4-1)
- 0.1234 / 0.10 = 1.23 (scale=2; 4-2)
- 0.1234 / 0.100 = 1.2 (scale=1; 4-3)
- 0.1234 / 0.1000 = 1 (scale=0; 4-4)

My first thought on this is rather unintuitive.
The higher the scale of the divisor the more "inaccurate" the result.
Is there another way to think about  this?

**Comment by bdice** [comment]
> Does this mean:
> - 0.1234 / 0.1 = 1.234 (scale=3; 4-1)
> - 0.1234 / 0.10 = 1.23 (scale=2; 4-2)
> - 0.1234 / 0.100 = 1.2 (scale=1; 4-3)
> - 0.1234 / 0.1000 = 1 (scale=0; 4-4)

Yes, the natural way to think about this is with the underlying integer representations. In [C integer division](https://godbolt.org/z/M6fnKq9hc) or [Rust integer division](https://godbolt.org/z/Md98zzWP6), we get what [Python calls "floor division"](https://godbolt.org/z/jj9fhY5E3). In that situation, if we drop the scales, we see how the integer representations naturally lead to this result.
- 1234 / 1 = 1234
- 1234 / 10 = 123
- 1234 / 100 = 12
- 1234 / 1000 = 1

Then you handle the scales accordingly, as `s1 - s2`.

As above with multiplication, there are downsides if the implementation chooses to do anything else but truncate according to the rules of integer division. If the implementation chooses to keep extra precision:
- You have to scale up the dividend's integer representation, like 1234 to 12340000. Downscaling the divisor could truncate the divisor and lead to incorrect results.
- That incurs an extra multiplication which can cause overflow (which could be detected by analyzing the values but again, this increases the computational cost)

If we choose to drop precision and move away from the existing behavior, the solution I would recommend is to accept truncation as a default behavior. Do integer division here, and use `s1 - s2` as the result scale. We can definitely improve the gap in education to address the downsides here: we should teach how to cast inputs so that division is safe against truncation. Users can increase the scale of the dividend by the scale of the divisor, if the possibility of overflow is preferred over truncation. Here's a sample implementation (this assumes the proposed behavior change of decimal division returning scale `s1 - s2`):
```python
def divide_without_truncation(s1, s2):
    """Divide decimal series s1 and s2 without integer truncation.

    Upscales s1 by the scale of s2, to ensure the result has the same scale as s1.
    """
    return s1.cast(pl.Decimal(scale=s1.dtype.scale + s2.dtype.scale)) / s2
```

It is important for operations like multiply and divide to do minimal math (no extra divisions for rescaling) and avoid loss of data to overflow. We ended up following this path in libcudf because it was the best way to let the user or calling library have full control over the conventions. It is easier to implement a different convention for how scales behave in a downstream library if the upstream implementation can round-trip through various math operations with minimal risk of overflow and does no extra rescaling work. It is always possible for a caller to scale up the inputs to get a desired output. It is harder to undo a convention like `max(s1, s2)` if the desired outcome is different (especially if it requires undoing a rescale that was already done by the implementation).

**Comment by hugopendlebury** [comment]
Would there still be a way to specify the precision and scale when writing to an external format such as a parquet file ?

We use polars to process weather data and save the resulting data into a parquet file which is then loaded into databricks.
Some of the loads have been large backfills of CDS data at a global level for multiple parameters (3 trillion rows) and HRES data (8 trillion rows)
Obviously databricks is a expensive vendor product however we found making a few simple changes such setting values to pl.Decimal(precision=15, scale=6) could result in significantly smaller parquet files.
It would be a shame if as a result of this change we had to drop down to pyarrow to perform some of the final steps.

**Comment by coastalwhite** [comment]
> Would there still be a way to specify the precision and scale when writing to an external format such as a parquet file ?

I do indeed think this is a prerequisite. #17418 should be prioritized if this becomes something.

**Comment by orlp** [comment]
@bdice I don't agree with a lot of your concerns around keeping the scale fixed on multiplication/division.

It is up to us to implement it correctly without overflow, assuming the result fits. But we can do that, even if it costs extra operations. Since the scales is limited to <= 38 we can also make a lookup table with efficient multipliers to replace division with multiplication.

> It is important for operations like multiply and divide to do minimal math (no extra divisions for rescaling) and avoid loss of data to overflow.

Why? This approach just shoves the problem to the user, for example you literally can't even multiply two scale-20 integers in this approach with Decimal128 - the scale itself overflows. Ultimately I believe the multiplication/division performance will be more than acceptable regardless, so I think we should focus on what the user actually wants.

**Comment by bdice** [comment]
> This approach just shoves the problem to the user

Another way to approach this is to have two layers of API. The low level API does the minimal thing which may be desired by library developers or power users who need to match another library’s conventions without risk of overflow (as we chose to do in libcudf), and this also exposes a no-overhead operation because no extra rescaling is performed. Then, a high-level (user facing) API can perform rescaling to adjust to a specific convention. This way users “get what they want” in most operations, but there is an escape route for people building tools or high-performance libraries on top of Polars that want the more fundamental operation (where the caller may have additional metadata, for example, and the case of overflowing the scale is not a concern).

**Comment by orlp** [comment]
@bdice If you don't do any rescaling then you're left with just a `Int128`. Which is in the process of being fully supported/exposed.

**Comment by TheLostLambda** [comment]
Likely a bit of a silly question from me not being familiar with the internals here, but is there a reason we're not using something like the Decimal type defined by the `rust_decimal` type which seems to be massively popular according to crates.io?

It's what's I've been using in all of my other code, and I sorta figured I'd be able to use a similar no-fuss decimal type in Polars: https://docs.rs/rust_decimal/latest/rust_decimal/

**Comment by orlp** [comment]
@TheLostLambda There is a wide variety of reasons we don't. The most primary reason is that we'd like to have full control over both data representation and implementation. For example, the `Decimal` in that crate wastes 32 bits on metadata for each `Decimal`, whereas we know we have the same metadata for an entire column, so that does not have to be repeated for each value.

**Comment by jherrmannNetfonds** [comment]
Hi, we use polars for financial transactions. Our main goal for using decimals is to have a predictable behavior (consistent rounding when necessary) when doing mathematical operation and a fixed output scale. 
For example if we have a portfolio of Stocks and calculate their sum (scale 2) we want `0.3+0.3+0.3` to be `0.90` not `0.8999999999999999` as with a floating point operation.
Our input values have different scales, but the desired output is always fixed scale (may differ per column)
With the current implementation the user experience is really bad (especially for junior devs who don't know what "decimal" is) since we quite often hit the s1+s2 case on multiplication. To make it currently work, you have to scatter casts everywhere which is non intuitive and makes the code highly unreadable.
